### PR TITLE
Implemented PIVOT/UNPIVOT Redshift + Fixed Snowflake Bug + Standardized PIVOT/UNPIVOT Parsing

### LINKED ISSUE #2875: PIVOT and UNPIVOT redshift
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Redshift recently added support for PIVOT and UNPIVOT, these are not recognised by SQLFluff and so result in a parsing error. 

https://aws.amazon.com/about-aws/whats-new/2022/02/amazon-redshift-support-pivot-unpivot-sql-operators/

### Use case

Redshift dialect support for PIVOT and UNPIVOT

### Dialect

Redshfit 

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #2877: `PIVOT` mistakenly parsed as an `alias_expression` after a subquery
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When parsing a query that contains a subquery in the FROM statement, followed by a PIVOT, sqlfluff recognizes the PIVOT as an alias.

Example query:
```sql
SELECT *
FROM (SELECT * FROM my_tbl)
PIVOT (min(f_val) FOR f_id IN (1, 2)) AS f (a, b);
```

### Expected Behaviour

Parse `PIVOT` as a `from_pivot_expression`

### Observed Behaviour

Parses `PIVOT` as an `alias_expression`.

Using the following minimal example:

```bash
echo "SELECT * FROM (SELECT * FROM my_tbl) PIVOT (min(f_val) FOR f_id IN (1, 2)) AS f (a, b);" | sqlfluff parse --dialect=snowflake -
```

I get the following error

```
==== parsing violations ====
L:   1 | P:  44 |  PRS | Line 1, Position 44: Found unparsable section: '(min(f_val) FOR
                       | f_id IN (1, 2)) AS f (a,...'
```

Though, really, the main offender is in the parse tree.

```
[L:  1, P: 38]      |                        alias_expression:                                                                                                                                                    
[L:  1, P: 38]      |                            identifier:                       'PIVOT' 
```

### How to reproduce

```bash
echo "SELECT * FROM (SELECT * FROM my_tbl) PIVOT (min(f_val) FOR f_id IN (1, 2)) AS f (a, b);" | sqlfluff parse --dialect=snowflake -
```

### Dialect

snowflake

### Version

0.11.1

### Configuration
```
[sqlfluff]
dialect = snowflake
; sqlfluff rules: https://docs.sqlfluff.com/en/stable/rules.html
rules = L001,
        ; Avoid trailing whitespace
        L002,
        ; Avoid mixing tabs and spaces
        L005,
        ; Avoid space before a comma
        L006,
        ; Use a single space around an operator
        L007,
        ; Use a newline before an operator, not after
        L008,
        ; Use a single space after a comma
        L009,
        ; Use a trailing newline at end of file
        L010,
        ; Use capital letters for keywords
        L015,
        ; Avoid parentheses after DISTINCT
        L020,
        ; Use unique aliases within a clause
        L021,
        ; Avoid ambiguous DISTINCT with GROUP BY
        L023,
        ; Use a single space after AS in a WITH clause
        L040,
        ; Use uppercase for TRUE, FALSE, NULL
        L041,
        ; Use DISTINCT and other SELECT modifiers on the same line as the clause
        L043,
        ; Avoid when/then/else statements returning boolean values
        L047,
        ; Use COUNT(*), not COUNT(0) or COUNT(1)
        Picnic_L001,
        ; Use consistent aliasing styles
        Picnic_L002,
        ; Use explicit aliases
        Picnic_L003,
        ; Use a single space around a casting operator
        Picnic_L004,
        ; Do not read from DV deprecated tables
        Picnic_L005,
        ; Avoid pattern IFF(x, TRUE, FALSE)
```

[sqlfluff:rules:L010]
capitalisation_policy = upper

[sqlfluff:rules:L040]
capitalisation_policy = upper

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by PLBMR**
### Brief summary of the change made

Fixes #2875 .
Fixes #2877 .

- **Implements `PIVOT/UNPIVOT` expressions for `redshift`**. This involved creating a `FromPivotExpressionSegment` and a `FromUnpivotExpressionSegment` in the `redshift` Dialect, which is inspired by the implementation of `PIVOT/UNPIVOT`s in the `snowflake` dialect. I introduced the following test cases to prove that this feature works for the `redshift` Dialect:
  - `test/fixtures/dialects/redshift/redshift_pivot.sql`
  - `test/fixtures/dialects/redshift/redshift_pivot.yml`
  - `test/fixtures/dialects/redshift/redshift_unpivot.sql`
  - `test/fixtures/dialects/redshift/redshift_unpivot.yml`
- Fixed the `alias_expression` bug surfaced in #2877 . I did this by adding `JoinLikeClauseGrammar` to the list of excluded aliases in the `FromExpressionElementSegment` in the `ansi` dialect.
- Refactored several `PIVOT/UNPIVOT` implementations in our dialect set to be simpler and more standardized. These changes include:
  - Updating the `JoinLikeClauseGrammar` in the `snowflake` dialect to end with an optional `AliasExpressionSegment` instead of an optional `TableAliasExpressionSegment`. This is because the implementation of `TableAliasExpressionSegment` was effectively redundant to `AliasExpressionSegment`. Consequently, I deprecated the `TableAliasExpressionSegment` segment in the `snowflake` dialect, since it was used in no other segments nor grammars.
  - Folded the `parse_grammar`s of `FromPivotExpressionSegment` and `FromUnpivotExpressionSegment` into their respective `match_grammar`s. The distinction between the `match_grammar` and `parse_grammar` in these segments were not useful since the original `match_grammar` did not indicate any relevant terminators. Thus, to align it with the implementation in the `redshift` dialect, I set the `match_grammar` in each of these segments to the original `parse_grammar` implementation, and deprecated their customized `parse_grammar` attributes.
  - ~Restructured `PIVOT/UNPIVOT` parsing in the `tsql` dialect to align with the structure of `PIVOT/UNPIVOT` parsing in the `snowflake` and `redshift` dialects. This will ensure that when we implement `PIVOT/UNPIVOT` expression parsing in future dialects, we have a standard implementation approach that is exemplified in other dialects.~ (reverted this change)

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`. **(N/A for this change)**
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`). **(DONE)**
    - `test/fixtures/dialects/redshift/redshift_pivot.sql`
    - `test/fixtures/dialects/redshift/redshift_pivot.yml`
    - `test/fixtures/dialects/redshift/redshift_unpivot.sql`
    - `test/fixtures/dialects/redshift/redshift_unpivot.yml`
    - `test/fixtures/dialects/snowflake/snowflake_pivot.yml`
  - Full autofix test cases in `test/fixtures/linter/autofix`. **(N/A for this change)**
  - Other. **(N/A for this change)**
- Added appropriate documentation for the change. **(N/A for this change)**
- Created GitHub issues for any relevant followup/future enhancements if appropriate. **(DONE)**
  - `tsql` dialect follow-up is raised in #2890.

### CODE REVIEW

**Comment by PLBMR** [comment]
Clearly, my changes to the `tsql` dialect have regressed some of the expected behavior in `L028`. I'll remove changes to `tsql` from this PR, and open up an issue for addressing the desired `tsql` changes later.

All other changes in this PR will remain intact.

**Comment by PLBMR** [comment]
T-SQL changes reverted. Snowflake and Redshift changes remain intact.

**Comment by PLBMR** [comment]
Surfaced desired restructuring for the `tsql` dialect in #2890 .

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 1]
```diff
@@ -222,6 +232,98 @@ class ColumnReferenceSegment(ObjectReferenceSegment):  # type: ignore
...
+        # Dedent for the indent in the select clause.
+        # It's here so that it can come AFTER any whitespace.
+        Dedent,
+        Ref("IntoClauseSegment", optional=True),
+        Ref("FromClauseSegment", optional=True),
+        Ref("WhereClauseSegment", optional=True),
+        Ref("GroupByClauseSegment", optional=True),
+        Ref("HavingClauseSegment", optional=True),
+        Ref("OverlapsClauseSegment", optional=True),
+    )
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_redshift.py:256]: I don't see the PIVOT/UNPIVOT additions. Is this needed?

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 2]
```diff
@@ -384,7 +384,7 @@
             Ref("SamplingExpressionSegment"),
             min_times=1,
         ),
-        Ref("TableAliasExpressionSegment", optional=True),
+        Ref("AliasExpressionSegment", optional=True),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:387]: Can we make this same change in BigQuery?

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 3]
```diff
@@ -871,21 +871,6 @@ class WithinGroupClauseSegment(BaseSegment):
     )
 
 
-@snowflake_dialect.segment()
-class TableAliasExpressionSegment(BaseSegment):
-    """A reference to an object with an `AS` clause, optionally with column aliasing."""
-
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_snowflake.py:877]: Again this can be removed from BigQuery

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Looks good. Nice clean up. Few small questions.

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 1 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_redshift.py:256]: Good callout. I think this is a vestigial of a previous implementation path. I'll deprecate this dialect replacement.

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 2 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_snowflake.py:387]: can do!

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 3 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_snowflake.py:877]: will do!

File: src/sqlfluff/dialects/dialect_redshift.py (hunk 1 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_redshift.py]: done!

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 3 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_snowflake.py:877]: done!

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 2 above)
Comment (PLBMR) [reply src/sqlfluff/dialects/dialect_snowflake.py:387]: done!

**Review by tunetheweb — APPROVED** [review]
Great work!

**Pull request merged**